[quant][pt2] Add Conv + BN fusion for prepare QAT

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98568

**Summary:** This commit adds the `prepare_qat_pt2e` API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the convert flow needs to swap the new subgraph
with another one that merges the batchnorm stats back into conv.

The Conv + BN fusion is implemented in the following steps:

1. Annotate all nodes in the pattern `[conv - bn - getitem]`

2. Match and replace this pattern with the fused QAT pattern
   (note that this is a larger subgraph than the original one)

3. Copy over metadata from the original nodes to the
   corresponding nodes in the new subgraph, to ensure the
   stack traces and dtype annotations are preserved

4. Prepare will insert fake quantizes in the right places
   based on the annotations

**Test Plan:**
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

**Reviewers:** jerryzh168, kimishpatel, yanboliang

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

andrewor14 committed:
[quant][pt2] Add Conv + BN fusion for prepare QAT

Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

yanboliang commented on torch/_dynamo/skipfiles.py:
```diff
@@ -147,6 +147,12 @@ def _module_dir(m: types.ModuleType):
+FILENAME_ALLOWLIST |= {
+    _module_dir(torch) + "ao/quantization/_pt2e/qat_utils.py",
+}
+
```
I would suggest you put this under:
https://github.com/pytorch/pytorch/blob/ebd4c165ffa586491dac51c555b8bffd1f844b5a/torch/_dynamo/skipfiles.py#L242-L245
This is where we define the rules if tracing through torch internal code.

andrewor14 replied:
Looks like `allow_torch` is False when I call `torchdynamo.export` however, so we'll skip this case:
https://github.com/pytorch/pytorch/blob/01a7f24f7367f6d13ffb63f69cc6523583187620/torch/_dynamo/skipfiles.py#L199-L200
How can I make it work?

andrewor14 replied:
Sounds good. However, looks like `allow_torch=False` when I call `torchdynamo.export`, so it'll skip this case:
https://github.com/pytorch/pytorch/blob/01a7f24f7367f6d13ffb63f69cc6523583187620/torch/_dynamo/skipfiles.py#L199-L200
How can I make it work?

yanboliang replied:
Oh, you can append your module here to allow inline/tracing:
https://github.com/pytorch/pytorch/blob/95621b3c2e8bc0f1d7f88fb0bcb3e3e56a0ff961/torch/_dynamo/config.py#L126-L135

andrewor14 replied:
Hmm, this didn't seem to work. I tried adding `torch.ao.quantization` here and I can't export modules defined in this namespace. The other problem is I can't specify a tighter namespace (e.g. `torch.ao.quantization.pt2e.qat_utils`) due to circular imports.

yanboliang replied:
Hmm, ```torch.ao.nn``` is here, so generally if you want the PyTorch ao builtin modules can be inlined/traced by dynamo, it should be there. Do you think it makes sense to move your module under ```torch.ao.nn```? This would make the structure more clear and easy to maintain.

andrewor14 replied:
I tried moving it to `torch.ao.nn` and I'm getting the same error when I export:
```
AssertionError: Failed to produce a graph during tracing. Tracing through 'f' must produce a single graph.
```
I feel `skipfiles_inline_module_allowlist` isn't working as we expect here.

yanboliang replied:
Oh, that indicates a circular imports during export. I think that's a bug if ```torch.ao.nn``` doesn't work, I'll check it and circle back to you.

yanboliang replied:
@andrewor14 Sorry for late response, was really busy last week, but finally figured it out.

I can reproduce your issue with https://gist.github.com/yanboliang/dcfa8c0e7f4249a2961630b8e98045d0.
Look at this repro: it passed with dynamo, but failed with export. The reason is export doesn't allowed None graph, but dynamo is allowed.
Why it produced None graph? because we skipped the trace of these modules. We put a ```call_module``` op in the FX graph at dynamo level and hand it over to AOT autograd to unroll/flatten these __builtin__ modules, as we can ensure they are safely traced by AOT autograd and make dynamo level simple to handle this case.
However, you are trying to allowed dynamo to trace through the modules under ```torch.ao.nn.*```, can you explain a bit more on why you have to do this?
Take a step back, if there is a strong reason you must to do this, I think the way is follow what we did for ```_prims``` and ```_refs``` at:
https://github.com/pytorch/pytorch/blob/be0b12ece576c86c5f059d15a64dcba0eb886ddd/torch/_dynamo/skipfiles.py#L142-L150

For the above repro I have, simply adding the following lines can make dynamo trace through the ```ao``` Linear module.
```
FILENAME_ALLOWLIST |= {torch.ao.nn.qat.modules.linear.__file__,}
```
I think it makes sense to trace through ```_prims``` and ```_refs```, but I didn't see the benefit of tracing ```ao``` builtin modules, can you elaborate more?

andrewor14 replied:
As discussed offline, there doesn't seem to be a good way to work around the circular import issue. To unblock, we will land this change as is (maybe add some comments to explain). In the future, we should see if we can do something else other than expressing the file name through a string.

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


Summary: This commit adds the prepare_qat_pt2e API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the following things are still missing and
will be added in future PRs:

1) Weight fake quantize after scaled weight
2) Convert flow needs to swap the new subgraph with another
   one that merges the batchnorm stats back into conv

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/_quantize_pt2e.py:
```diff
@@ -70,7 +71,7 @@ def prepare_pt2e_quantizer(
         model,
         quantizer,
-        False,  # is_qat
+        is_qat,
```
wait, how does this work? in L70 we already run _fuse_conv_bn_ for PTQ

jerryzh168 replied:
I guess that is not run beacuse batchnorm has a different form in training

andrewor14 replied:
Right I meant to put `fuse_conv_bn` under `if not qat` but it got lost in refactoring

andrewor14 replied:
Right I meant to put `fuse_conv_bn` under `if not qat` but it got lost in refactoring. I'll separate prepare_qat out

andrewor14 replied:
Oh yeah I meant to put `fuse_conv_bn` under `if not is_qat` but it got lost in refactoring

jerryzh168 commented on torch/ao/quantization/_quantize_pt2e.py:
```diff
@@ -79,6 +80,13 @@ def prepare_pt2e_quantizer(
+    model: GraphModule,
+    quantizer: Quantizer,
+):
+    return prepare_pt2e_quantizer(model, quantizer, is_qat=True)
```
I feel we should probably not call the PTQ version of this api, but instead just do the same thing as `prepare_pt2e_quantizer` here but change `_fuse_conv_bn_` to `_fuse_conv_bn_qat`

andrewor14 replied:
Sounds good. However, we will need to do `_fuse_conv_bn_qat` between annotate and prepare. Should I move annotate out of prepare?

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


**Summary:** This commit adds the `prepare_qat_pt2e` API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the convert flow needs to swap the new subgraph
with another one that merges the batchnorm stats back into conv.

The Conv + BN fusion is implemented in the following steps:

1. Annotate all nodes in the pattern `[conv - bn - getitem]`

2. Match and replace this pattern with the fused QAT pattern
   (note that this is a larger subgraph than the original one)

3. Copy over metadata from the original nodes to the
   corresponding nodes in the new subgraph, to ensure the
   stack traces and dtype annotations are preserved

4. Prepare will insert fake quantizes in the right places
   based on the annotations

**Test Plan:**
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

**Reviewers:** jerryzh168, kimishpatel, yanboliang

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


**Summary:** This commit adds the `prepare_qat_pt2e` API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the convert flow needs to swap the new subgraph
with another one that merges the batchnorm stats back into conv.

The Conv + BN fusion is implemented in the following steps:

1. Annotate all nodes in the pattern `[conv - bn - getitem]`

2. Match and replace this pattern with the fused QAT pattern
   (note that this is a larger subgraph than the original one)

3. Copy over metadata from the original nodes to the
   corresponding nodes in the new subgraph, to ensure the
   stack traces and dtype annotations are preserved

4. Prepare will insert fake quantizes in the right places
   based on the annotations

**Test Plan:**
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

**Reviewers:** jerryzh168, kimishpatel, yanboliang

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


**Summary:** This commit adds the `prepare_qat_pt2e` API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the convert flow needs to swap the new subgraph
with another one that merges the batchnorm stats back into conv.

The Conv + BN fusion is implemented in the following steps:

1. Annotate all nodes in the pattern `[conv - bn - getitem]`

2. Match and replace this pattern with the fused QAT pattern
   (note that this is a larger subgraph than the original one)

3. Copy over metadata from the original nodes to the
   corresponding nodes in the new subgraph, to ensure the
   stack traces and dtype annotations are preserved

4. Prepare will insert fake quantizes in the right places
   based on the annotations

**Test Plan:**
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

**Reviewers:** jerryzh168, kimishpatel, yanboliang

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_quantize_pt2e.py:
```diff
@@ -20,6 +20,7 @@
     QConfigMapping,
     observer,
 )
+from torch.ao.quantization.fx.prepare import _insert_observer
```
why do we need this? this is an internal function that we may remove or change, I don't think it should be used in tests

jerryzh168 commented on test/quantization/fx/test_quantize_pt2e.py:
```diff
@@ -299,6 +314,61 @@ def forward(self, x):
+        m = prepare_qat_pt2e_quantizer(m, quantizer)
+        m(*example_inputs)
+
+        # Verify that the subgraph was replaced successfully with fake quantizes inserted
```
can we just check the result model graph directly

kimishpatel replied:
yeah, that seems more robust?

andrewor14 replied:
Sure. I will just match it manually

jerryzh168 commented on torch/ao/quantization/_pt2e/qat_utils.py:
```diff
@@ -0,0 +1,137 @@
+    example_inputs = _conv_bn_pattern_example_inputs
+    match_pattern = _get_aten_graph_module(_conv_bn_pattern, example_inputs)
+    replacement_pattern = _get_aten_graph_module(_fused_qat_conv_bn_pattern, example_inputs)
+    match_and_replacement = _replace_pattern(m, match_pattern, replacement_pattern)
```
I feel we should not be calling internal functions for subgraph rewriter as well..

kimishpatel replied:
Agreed!

kimishpatel replied:
Maybe file a TODO here if you want to fix in the next PR

andrewor14 replied:
Sounds good. Yes the public facing one doesn't return replacement information right now

jerryzh168 commented on torch/ao/quantization/_pt2e/qat_utils.py:
```diff
@@ -0,0 +1,137 @@
+        while replacement_conv_node is None or replacement_bn_node is None:
+            if n.target == torch.ops.aten.convolution.default:
+                replacement_conv_node = n
+            if n.target == torch.ops.aten._native_batch_norm_legit.default:
```
is this op also used in training? I thought this is just for eval?

andrewor14 replied:
Looks like it's used for training as well

jerryzh168 commented on torch/ao/quantization/_pt2e/qat_utils.py:
```diff
@@ -0,0 +1,137 @@
+        assert len(mr.replacements) == 1, "expected only one replacement node"
+        replacement_conv_node = None
+        replacement_bn_node = None
+        replacement_anchor = mr.replacements[0]
```
nit: `replacement_anchor` --> `replacement_getitem` might be a bit clearer and also add a comment that this is the getitem for the output of bn will be helpful as well

jerryzh168 commented on torch/ao/quantization/_pt2e/utils.py:
```diff
@@ -127,3 +128,15 @@ def _rearrange_weight_observer_for_decomposed_linear(
     model.graph.lint()
     model.recompile()
+
+def _build_node_name_to_scope(model: GraphModule) -> Dict[str, Tuple[str, type]]:
```
nit: build --> get might be clearer I feel

kimishpatel replied:
@jerryzh168 is this actively used? If this is set to empty dict, would the workflow still work?

andrewor14 replied:
Let's do that separately. Here I'm just refactoring to reduce duplicate code

jerryzh168 commented on torch/ao/quantization/_quantize_pt2e.py:
```diff
@@ -53,27 +47,32 @@ def prepare_pt2e_quantizer(
+    model = prepare(model, node_name_to_scope, is_qat=False)
+    # TODO: remove hack when we have better support for pattern matching
+    # move around the observer for addmm
+    _rearrange_weight_observer_for_decomposed_linear(model)
```
don't need to do it in this PR but I think we can probably remove this now

kimishpatel replied:
This is also something I would like to not have to worry about. So we either pattern match the subgraph or use module level info

jerryzh168 commented on torch/ao/quantization/_pt2e/qat_utils.py:
```diff
@@ -0,0 +1,137 @@
+            n = n.args[0]
+
+        # Copy over metadata for conv node, bn node, conv args, bn args, and bn users
+        for match_pattern_node, original_node in mr.nodes_map.items():
```
do need to copy over meta for args? I think arg nodes are not really being replaced by subgraph rewriter since they are the input of the pattern?

andrewor14 replied:
Good catch. This was leftover from when we used modules to express the patterns. I just verified that we don't need to copy for the args

jerryzh168 reviewed (commented):
looks great overall, had some comments inline

kimishpatel commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -285,6 +332,56 @@ def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
+            return
+        getitem_node = node
+        bn_node = getitem_node.args[0]
+        assert isinstance(bn_node, Node)
```
What else can this be? If so, would it be better to return early instead of assert? Also instead of assert can we raise exception?

andrewor14 replied:
This is mainly just for mypy. I think it's fine to assert here. We do the same for all the other annotations in this file too

kimishpatel commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -285,6 +332,56 @@ def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
+        getitem_node = node
+        bn_node = getitem_node.args[0]
+        assert isinstance(bn_node, Node)
+        if bn_node.op != "call_function" or bn_node.target != torch.ops.aten._native_batch_norm_legit.default:
```
curious, is this is the only op that would show up as bn?

andrewor14 replied:
It seems so for now. If we find others we can add in future PRs

kimishpatel commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -285,6 +332,56 @@ def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
+            "bias_index": 2,
+            "_annotated": True,
+        }
+        bn_node.meta["target_dtype_info"] = {
+            "input_act_obs_or_fq_ctr": _get_default_obs_or_fq_ctr(),
+            "output_act_obs_or_fq_ctr": _get_default_obs_or_fq_ctr(),
+            "_annotated": True,
+        }
```
I am assuming we wont need this after @jerryzh168's stuff where not all nodes need annotation?

jerryzh168 replied:
yes, these can be removed after https://github.com/pytorch/pytorch/pull/99220 is landed

kimishpatel commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -285,6 +332,56 @@ def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
+            "_annotated": True,
+        }
+        getitem_node.meta["target_dtype_info"] = {
+            "input_act_obs_or_fq_ctr": _get_default_obs_or_fq_ctr(),
```
Same here. In fact with this PR, https://github.com/pytorch/pytorch/pull/99071, you can just say None for input.

andrewor14 replied:
Ok, let's do this later since none of the other annotations are updated to reflect that yet

kimishpatel commented on torch/ao/quantization/_pt2e/qat_utils.py:
```diff
@@ -0,0 +1,137 @@
+
+
+# Example inputs for both `_conv_bn_pattern` and `_fused_qat_conv_bn_pattern`
+_conv_bn_pattern_example_inputs = (
```
we should call this conv2d pattern, right?

kimishpatel commented on torch/ao/quantization/_pt2e/qat_utils.py:
```diff
@@ -0,0 +1,137 @@
+    bn_running_var: torch.Tensor,
+) -> torch.Tensor:
+    """
+    Approximated method to fuse conv and bn. It requires only one forward pass.
```
Is this copied over from quantizable ConvBN? If so can you reference the code pointer.

andrewor14 replied:
Yes it's referenced two lines below this

kimishpatel commented on torch/ao/quantization/_pt2e/qat_utils.py:
```diff
@@ -0,0 +1,137 @@
+
+    # Copy over metadata from original subgraph
+    # This ensures the stack traces and annotations are preserved in the new subgraph
+    for mr in match_and_replacement:
```
Also for this if we can have subgraph rewriter support it, it might be great. We did this for TS variant here https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/passes/subgraph_rewrite.cpp#L56, but probably harder to do here since you dont have access to internal nodes of the pattern

andrewor14 replied:
Added a TODO

kimishpatel commented on torch/ao/quantization/_pt2e/qat_utils.py:
```diff
@@ -0,0 +1,137 @@
+                assert zeros_like_node.target == torch.ops.aten.zeros_like.default
+                mul_node.args[0].meta = mr.nodes_map[weight].meta
+                zeros_like_node.args[0].meta = mr.nodes_map[bias].meta
+                replacement_conv_node.meta = original_node.meta
```
we should be able to have subgraph rewriter take care of this logic too.

kimishpatel commented on test/quantization/fx/test_quantize_pt2e.py:
```diff
@@ -299,6 +314,61 @@ def forward(self, x):
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.conv = torch.nn.Conv2d(3, 3, 3)
```
can you also add a test where we have conv + conv + bn?

andrewor14 replied:
Ok, let's do it in a future PR

kimishpatel commented on test/quantization/fx/test_quantize_pt2e.py:
```diff
@@ -299,6 +314,61 @@ def forward(self, x):
+        check_pattern.graph.eliminate_dead_code()
+        check_pattern.recompile()
+        matcher = SubgraphMatcher(check_pattern.graph)
+        self.assertEqual(len(matcher.match(m.graph)), 1)
```
Also can you add a check for preserving metadata? or on in the follow up PR?

andrewor14 replied:
Added a TODO

kimishpatel reviewed (commented):
Left some comments. Address the essential ones. For the rest you can do in follow up PR

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


**Summary:** This commit adds the `prepare_qat_pt2e` API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the convert flow needs to swap the new subgraph
with another one that merges the batchnorm stats back into conv.

The Conv + BN fusion is implemented in the following steps:

1. Annotate all nodes in the pattern `[conv - bn - getitem]`

2. Match and replace this pattern with the fused QAT pattern
   (note that this is a larger subgraph than the original one)

3. Copy over metadata from the original nodes to the
   corresponding nodes in the new subgraph, to ensure the
   stack traces and dtype annotations are preserved

4. Prepare will insert fake quantizes in the right places
   based on the annotations

**Test Plan:**
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

**Reviewers:** jerryzh168, kimishpatel, yanboliang

[ghstack-poisoned]

kimishpatel commented:
@andrewor14 let me know when its ready for re-review

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


**Summary:** This commit adds the `prepare_qat_pt2e` API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the convert flow needs to swap the new subgraph
with another one that merges the batchnorm stats back into conv.

The Conv + BN fusion is implemented in the following steps:

1. Annotate all nodes in the pattern `[conv - bn - getitem]`

2. Match and replace this pattern with the fused QAT pattern
   (note that this is a larger subgraph than the original one)

3. Copy over metadata from the original nodes to the
   corresponding nodes in the new subgraph, to ensure the
   stack traces and dtype annotations are preserved

4. Prepare will insert fake quantizes in the right places
   based on the annotations

**Test Plan:**
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

**Reviewers:** jerryzh168, kimishpatel, yanboliang

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


**Summary:** This commit adds the `prepare_qat_pt2e` API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the convert flow needs to swap the new subgraph
with another one that merges the batchnorm stats back into conv.

The Conv + BN fusion is implemented in the following steps:

1. Annotate all nodes in the pattern `[conv - bn - getitem]`

2. Match and replace this pattern with the fused QAT pattern
   (note that this is a larger subgraph than the original one)

3. Copy over metadata from the original nodes to the
   corresponding nodes in the new subgraph, to ensure the
   stack traces and dtype annotations are preserved

4. Prepare will insert fake quantizes in the right places
   based on the annotations

**Test Plan:**
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

**Reviewers:** jerryzh168, kimishpatel, yanboliang

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


**Summary:** This commit adds the `prepare_qat_pt2e` API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the convert flow needs to swap the new subgraph
with another one that merges the batchnorm stats back into conv.

The Conv + BN fusion is implemented in the following steps:

1. Annotate all nodes in the pattern `[conv - bn - getitem]`

2. Match and replace this pattern with the fused QAT pattern
   (note that this is a larger subgraph than the original one)

3. Copy over metadata from the original nodes to the
   corresponding nodes in the new subgraph, to ensure the
   stack traces and dtype annotations are preserved

4. Prepare will insert fake quantizes in the right places
   based on the annotations

**Test Plan:**
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

**Reviewers:** jerryzh168, kimishpatel, yanboliang

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


**Summary:** This commit adds the `prepare_qat_pt2e` API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the convert flow needs to swap the new subgraph
with another one that merges the batchnorm stats back into conv.

The Conv + BN fusion is implemented in the following steps:

1. Annotate all nodes in the pattern `[conv - bn - getitem]`

2. Match and replace this pattern with the fused QAT pattern
   (note that this is a larger subgraph than the original one)

3. Copy over metadata from the original nodes to the
   corresponding nodes in the new subgraph, to ensure the
   stack traces and dtype annotations are preserved

4. Prepare will insert fake quantizes in the right places
   based on the annotations

**Test Plan:**
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

**Reviewers:** jerryzh168, kimishpatel, yanboliang

[ghstack-poisoned]

kimishpatel reviewed (approved):
Looks good. Lets follow up on the todos or maybe file a task to keep track

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


**Summary:** This commit adds the `prepare_qat_pt2e` API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the convert flow needs to swap the new subgraph
with another one that merges the batchnorm stats back into conv.

The Conv + BN fusion is implemented in the following steps:

1. Annotate all nodes in the pattern `[conv - bn - getitem]`

2. Match and replace this pattern with the fused QAT pattern
   (note that this is a larger subgraph than the original one)

3. Copy over metadata from the original nodes to the
   corresponding nodes in the new subgraph, to ensure the
   stack traces and dtype annotations are preserved

4. Prepare will insert fake quantizes in the right places
   based on the annotations

**Test Plan:**
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

**Reviewers:** jerryzh168, kimishpatel, yanboliang

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


**Summary:** This commit adds the `prepare_qat_pt2e` API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the convert flow needs to swap the new subgraph
with another one that merges the batchnorm stats back into conv.

The Conv + BN fusion is implemented in the following steps:

1. Annotate all nodes in the pattern `[conv - bn - getitem]`

2. Match and replace this pattern with the fused QAT pattern
   (note that this is a larger subgraph than the original one)

3. Copy over metadata from the original nodes to the
   corresponding nodes in the new subgraph, to ensure the
   stack traces and dtype annotations are preserved

4. Prepare will insert fake quantizes in the right places
   based on the annotations

**Test Plan:**
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

**Reviewers:** jerryzh168, kimishpatel, yanboliang

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 commented:
@pytorchbot merge

jerryzh168 commented:
@andrewor14 might be better to import to fbcode and merge the phabricator diff, we have confirmed with Nikita that this is OK for our project

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


**Summary:** This commit adds the `prepare_qat_pt2e` API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the convert flow needs to swap the new subgraph
with another one that merges the batchnorm stats back into conv.

The Conv + BN fusion is implemented in the following steps:

1. Annotate all nodes in the pattern `[conv - bn - getitem]`

2. Match and replace this pattern with the fused QAT pattern
   (note that this is a larger subgraph than the original one)

3. Copy over metadata from the original nodes to the
   corresponding nodes in the new subgraph, to ensure the
   stack traces and dtype annotations are preserved

4. Prepare will insert fake quantizes in the right places
   based on the annotations

**Test Plan:**
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

**Reviewers:** jerryzh168, kimishpatel, yanboliang

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 commented:
I feel that wouldn't help with the merge conflicts here since it's conflicting on OSS only files. Also phabricator tests take longer to run so I feel that would delay merging this PR further.

andrewor14 commented:
I see. Ok, maybe we can discuss in the meeting on the best way to move forward.

andrewor14 committed:
Update on "[quant][pt2] Add Conv + BN fusion for prepare QAT"


**Summary:** This commit adds the `prepare_qat_pt2e` API and the
fusion logic for Conv + BN. We use the subgraph rewriter to
match and replace the pattern with the existing logic in
`nniqat.ConvBn2d`. Note this is not the end-to-end flow yet.
In particular, the convert flow needs to swap the new subgraph
with another one that merges the batchnorm stats back into conv.

The Conv + BN fusion is implemented in the following steps:

1. Annotate all nodes in the pattern `[conv - bn - getitem]`

2. Match and replace this pattern with the fused QAT pattern
   (note that this is a larger subgraph than the original one)

3. Copy over metadata from the original nodes to the
   corresponding nodes in the new subgraph, to ensure the
   stack traces and dtype annotations are preserved

4. Prepare will insert fake quantizes in the right places
   based on the annotations

**Test Plan:**
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion

**Reviewers:** jerryzh168, kimishpatel, yanboliang

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 commented:
@pytorchbot merge

kimishpatel commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -147,26 +164,46 @@ def _get_weight_obs_or_fq_ctr(quantization_config: Optional[QuantizationConfig])
+            )
+        elif quantization_spec.qscheme == torch.per_channel_symmetric:
+            return FusedMovingAvgObsFakeQuantize.with_args(
+                observer=MovingAveragePerChannelMinMaxObserver,
```
Andrew, why does fake quantize have observer type specified here for per channel, unlike observer? Can this not be refactored to just take per channel arg?

andrewor14 replied:
This is just following the existing implementation here: https://github.com/pytorch/pytorch/blob/0ebd3a78f483b7ea989276d2e29ef0ddce4ab153/torch/ao/quantization/fake_quantize.py#L465-L471

Not sure what you mean by "refactored to just take per channel arg"? You mean reducing the duplicate code in the two different qscheme cases?

kimishpatel replied:
> You mean reducing the duplicate code in the two different qscheme cases

Yes. Immediately this is just about factoring out observer type.

Longer term is having obsever just take per_channel arg rather than having having separate PerChannel observer
